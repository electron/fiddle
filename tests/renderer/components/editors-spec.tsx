import * as React from 'react';

import { mount, shallow } from 'enzyme';

import { EditorValues, MAIN_JS } from '../../../src/interfaces';
import { IpcEvents } from '../../../src/ipc-events';
import { Editors } from '../../../src/renderer/components/editors';
import * as content from '../../../src/renderer/content';
import { EditorMosaic } from '../../../src/renderer/editor-mosaic';
import { ipcRendererManager } from '../../../src/renderer/ipc';
import {
  AppMock,
  MonacoEditorMock,
  StateMock,
  createEditorValues,
} from '../../mocks/mocks';

jest.mock('../../../src/renderer/components/editor', () => ({
  Editor: () => 'Editor',
}));

describe('Editors component', () => {
  let app: AppMock;
  let monaco: any;
  let store: StateMock;
  let editorMosaic: EditorMosaic;
  let editorValues: EditorValues;

  beforeEach(() => {
    ({ app, monaco } = (window as any).ElectronFiddle);
    ({ state: store } = app);
    editorValues = createEditorValues();
    editorMosaic = new EditorMosaic();
    editorMosaic.set(editorValues);

    store.editorMosaic = editorMosaic as any;
  });

  it('renders', () => {
    const wrapper = mount(<Editors appState={store as any} />);
    wrapper.setState({ monaco });
    expect(wrapper).toMatchSnapshot();
  });

  it('does not execute command if not supported', () => {
    const wrapper = shallow(<Editors appState={store as any} />);
    const instance: any = wrapper.instance() as any;

    const editor = new MonacoEditorMock();
    const action = editor.getAction();
    action.isSupported.mockReturnValue(false);

    editorMosaic.focusedEditor = jest.fn().mockReturnValue(editor);

    instance.executeCommand('hello');
    expect(editor.getAction).toHaveBeenCalled();
    expect(action.isSupported).toHaveBeenCalled();
    expect(action.run).toHaveBeenCalledTimes(0);
  });

  describe('toggleEditorOption()', () => {
    const filename = MAIN_JS;

    it('handles an error', () => {
      const editor = new MonacoEditorMock();
      editorMosaic.addEditor(filename, editor as any);
      editor.updateOptions.mockImplementationOnce(() => {
        throw new Error('Bwap bwap');
      });

      const wrapper = shallow(<Editors appState={store as any} />);
      const instance: any = wrapper.instance() as any;

      expect(instance.toggleEditorOption('wordWrap')).toBe(false);
    });

    it('updates a setting', () => {
      const wrapper = shallow(<Editors appState={store as any} />);
      const instance: any = wrapper.instance() as any;

      const editor = new MonacoEditorMock();
      editorMosaic.addEditor(filename, editor as any);
      expect(instance.toggleEditorOption('wordWrap')).toBe(true);
      expect(editor.updateOptions).toHaveBeenCalledWith({
        minimap: { enabled: false },
        wordWrap: 'off',
      });
    });
  });

  it('renders a toolbar', () => {
    const wrapper = shallow(<Editors appState={store as any} />);
    const instance: any = wrapper.instance() as any;
    const toolbar = instance.renderToolbar({ title: MAIN_JS } as any, MAIN_JS);

    expect(toolbar).toMatchSnapshot();
  });

  it('onChange() updates the mosaic arrangement in the appState', () => {
    const wrapper = shallow(<Editors appState={store as any} />);
    const instance: any = wrapper.instance() as any;

    const arrangement = { testArrangement: true };
    instance.onChange(arrangement as any);
    expect(editorMosaic.mosaic).toStrictEqual(arrangement);
  });

  describe('IPC commands', () => {
    it('handles a MONACO_EXECUTE_COMMAND command', () => {
      shallow(<Editors appState={store as any} />);

      const editor = new MonacoEditorMock();
      const action = editor.getAction();
      action.isSupported.mockReturnValue(true);

      editorMosaic.focusedEditor = jest.fn().mockReturnValue(editor);

      ipcRendererManager.emit(IpcEvents.MONACO_EXECUTE_COMMAND, null, 'hello');
      expect(editor.getAction).toHaveBeenCalled();
      expect(action.isSupported).toHaveBeenCalled();
      expect(action.run).toHaveBeenCalled();
    });

    const fakeValues = { [MAIN_JS]: 'hi' } as const;

    it('handles an FS_NEW_FIDDLE command', async () => {
      let resolve: any;
      const replacePromise = new Promise((r) => {
        resolve = r;
      });

      // setup
      const getTemplateSpy = jest
        .spyOn(content, 'getTemplate')
        .mockImplementation(() => Promise.resolve(fakeValues));
      const replaceFiddleSpy = jest
        .spyOn(app, 'replaceFiddle')
        .mockImplementation(() => resolve());

      // invoke the call
      ipcRendererManager.emit(IpcEvents.FS_NEW_FIDDLE, null);
      await replacePromise;

      // check the results
      expect(getTemplateSpy).toHaveBeenCalledTimes(1);
      expect(replaceFiddleSpy).toHaveBeenCalledTimes(1);

      // cleanup
      getTemplateSpy.mockRestore();
      replaceFiddleSpy.mockRestore();
    });

    describe('SELECT_ALL_IN_EDITOR handler', () => {
      it('selects all in the focused editor', async () => {
        shallow(<Editors appState={store as any} />);

        const range = 'range';
        const editor = new MonacoEditorMock();
        const model = editor.getModel();
        model.getFullModelRange.mockReturnValue(range);
        editorMosaic.focusedEditor = jest.fn().mockReturnValue(editor);

        ipcRendererManager.emit(IpcEvents.SELECT_ALL_IN_EDITOR, null);

        await process.nextTick;
        expect(editor.setSelection).toHaveBeenCalledWith('range');
      });

      it('does not change selection if the selected editor has no model', async () => {
        shallow(<Editors appState={store as any} />);

        const editor = new MonacoEditorMock();
        delete (editor as any).model;
        editorMosaic.focusedEditor = jest.fn().mockReturnValue(editor);

        ipcRendererManager.emit(IpcEvents.SELECT_ALL_IN_EDITOR, null);

        await process.nextTick;
        expect(editor.getModel).toHaveBeenCalledTimes(1);
        expect(editor.setSelection).not.toHaveBeenCalled();
      });

      it('does not crash if there is no selected editor', () => {
        shallow(<Editors appState={store as any} />);
        editorMosaic.focusedEditor = jest.fn().mockReturnValue(null);
        ipcRendererManager.emit(IpcEvents.SELECT_ALL_IN_EDITOR, null);
      });
    });

    it('handles an FS_NEW_TEST command', async () => {
      // setup
      const getTestTemplateSpy = jest
        .spyOn(content, 'getTestTemplate')
        .mockImplementation(() => Promise.resolve(fakeValues));
      let replaceResolve: any;
      const replacePromise = new Promise((r) => {
        replaceResolve = r;
      });
      const replaceFiddleSpy = jest
        .spyOn(app, 'replaceFiddle')
        .mockImplementation(() => replaceResolve());

      // invoke the call
      ipcRendererManager.emit(IpcEvents.FS_NEW_TEST);
      await replacePromise;

      // check the results
      expect(getTestTemplateSpy).toHaveBeenCalled();
      expect(replaceFiddleSpy).toHaveBeenCalled();

      // cleanup
      getTestTemplateSpy.mockRestore();
      replaceFiddleSpy.mockRestore();
    });
    it('handles a SELECT_ALL_IN_EDITOR command', async () => {
      shallow(<Editors appState={store as any} />);

      const range = 'range';
      const editor = new MonacoEditorMock();
      editor.getModel().getFullModelRange.mockReturnValue(range);
      editorMosaic.focusedEditor = jest.fn().mockReturnValue(editor);

      ipcRendererManager.emit(IpcEvents.SELECT_ALL_IN_EDITOR, null);
      await process.nextTick;

      expect(editor.setSelection).toHaveBeenCalledWith(range);
    });

    it('handles the monaco editor option commands', () => {
      const id = MAIN_JS;
      const editor = new MonacoEditorMock();
      editorMosaic.addEditor(id, editor as any);

      shallow(<Editors appState={store as any} />);
      ipcRendererManager.emit(IpcEvents.MONACO_TOGGLE_OPTION, null, 'wordWrap');
      expect(editor.updateOptions).toHaveBeenCalled();
    });
  });

  describe('setFocused()', () => {
    it('sets the "focused" property', () => {
      const wrapper = shallow(<Editors appState={store as any} />);
      const instance: any = wrapper.instance() as any;
      const spy = jest.spyOn(instance, 'setState');

      const id = MAIN_JS;
      instance.setFocused(id);
      expect(spy).toHaveBeenCalledWith({ focused: id });
    });

    it('focus sidebar file', () => {
      const wrapper = shallow(<Editors appState={store as any} />);
      const instance: any = wrapper.instance() as any;
      const spy = jest.spyOn(instance, 'setState');

      const id = MAIN_JS;
      instance.setFocused(id);
      expect(spy).toHaveBeenCalledWith({ focused: id });
      expect(instance.props.appState.editorMosaic.focusedFile).toBe(id);
    });
  });
});
