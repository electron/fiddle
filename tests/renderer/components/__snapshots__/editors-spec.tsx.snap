// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Editors component renders 1`] = `
<div
  className="focused__undefined mosaic mosaic-drop-target"
>
  <div
    className="mosaic-root"
  >
    <div
      className="mosaic-tile"
      key="main.js"
      style={
        Object {
          "bottom": "50%",
          "left": "0%",
          "right": "50%",
          "top": "0%",
        }
      }
    >
      <div
        className="mosaic-window mosaic-drop-target main.js"
      >
        <div
          className="mosaic-window-toolbar draggable"
        >
          <div>
            <div>
              <h5>
                Main Process (main.js)
              </h5>
            </div>
            <div />
            <div
              className="mosaic-controls"
            >
              <button
                className="bp3-button bp3-small"
                onBlur={[Function]}
                onClick={[Function]}
                onKeyDown={[Function]}
                onKeyUp={[Function]}
                type="button"
              >
                <span
                  className="bp3-icon bp3-icon-maximize"
                  icon="maximize"
                >
                  <svg
                    data-icon="maximize"
                    height={16}
                    viewBox="0 0 16 16"
                    width={16}
                  >
                    <desc>
                      maximize
                    </desc>
                    <path
                      d="M5.99 8.99c-.28 0-.53.11-.71.29l-3.29 3.29v-1.59c0-.55-.45-1-1-1s-1 .45-1 1v4c0 .55.45 1 1 1h4c.55 0 1-.45 1-1s-.45-1-1-1H3.41L6.7 10.7a1.003 1.003 0 00-.71-1.71zm9-9h-4c-.55 0-1 .45-1 1s.45 1 1 1h1.59l-3.3 3.3a.99.99 0 00-.29.7 1.003 1.003 0 001.71.71l3.29-3.29V5c0 .55.45 1 1 1s1-.45 1-1V1c0-.56-.45-1.01-1-1.01z"
                      fillRule="evenodd"
                      key="0"
                    />
                  </svg>
                </span>
              </button>
              <button
                className="bp3-button bp3-small"
                onBlur={[Function]}
                onClick={[Function]}
                onKeyDown={[Function]}
                onKeyUp={[Function]}
                type="button"
              >
                <span
                  className="bp3-icon bp3-icon-cross"
                  icon="cross"
                >
                  <svg
                    data-icon="cross"
                    height={16}
                    viewBox="0 0 16 16"
                    width={16}
                  >
                    <desc>
                      cross
                    </desc>
                    <path
                      d="M9.41 8l3.29-3.29c.19-.18.3-.43.3-.71a1.003 1.003 0 00-1.71-.71L8 6.59l-3.29-3.3a1.003 1.003 0 00-1.42 1.42L6.59 8 3.3 11.29c-.19.18-.3.43-.3.71a1.003 1.003 0 001.71.71L8 9.41l3.29 3.29c.18.19.43.3.71.3a1.003 1.003 0 00.71-1.71L9.41 8z"
                      fillRule="evenodd"
                      key="0"
                    />
                  </svg>
                </span>
              </button>
            </div>
          </div>
        </div>
        <div
          className="mosaic-window-body"
        >
          Editor
        </div>
        <div
          className="mosaic-window-body-overlay"
          onClick={[Function]}
        />
        <div
          className="mosaic-window-additional-actions-bar"
        />
        <div
          className="mosaic-preview"
        >
          <div
            className="mosaic-window-toolbar"
          >
            <div
              className="mosaic-window-title"
            >
              Main Process (main.js)
            </div>
          </div>
          <div
            className="mosaic-window-body"
          >
            <h4>
              Main Process (main.js)
            </h4>
            <span
              className="bp3-icon bp3-icon-application"
              icon="application"
            >
              <svg
                data-icon="application"
                height={72}
                viewBox="0 0 20 20"
                width={72}
              >
                <desc>
                  application
                </desc>
                <path
                  d="M3.5 9h9c.28 0 .5-.22.5-.5s-.22-.5-.5-.5h-9c-.28 0-.5.22-.5.5s.22.5.5.5zm0 2h5c.28 0 .5-.22.5-.5s-.22-.5-.5-.5h-5c-.28 0-.5.22-.5.5s.22.5.5.5zM19 1H1c-.55 0-1 .45-1 1v16c0 .55.45 1 1 1h18c.55 0 1-.45 1-1V2c0-.55-.45-1-1-1zm-1 16H2V6h16v11zM3.5 13h7c.28 0 .5-.22.5-.5s-.22-.5-.5-.5h-7c-.28 0-.5.22-.5.5s.22.5.5.5z"
                  fillRule="evenodd"
                  key="0"
                />
              </svg>
            </span>
          </div>
        </div>
        <div
          className="drop-target-container"
        >
          <div
            className="drop-target top"
          />
          <div
            className="drop-target bottom"
          />
          <div
            className="drop-target left"
          />
          <div
            className="drop-target right"
          />
        </div>
      </div>
    </div>
    <div
      className="mosaic-split -column"
      onMouseDown={[Function]}
      style={
        Object {
          "bottom": "0%",
          "left": "0%",
          "right": "50%",
          "top": "50%",
        }
      }
    >
      <div
        className="mosaic-split-line"
      />
    </div>
    <div
      className="mosaic-tile"
      key="renderer.js"
      style={
        Object {
          "bottom": "0%",
          "left": "0%",
          "right": "50%",
          "top": "50%",
        }
      }
    >
      <div
        className="mosaic-window mosaic-drop-target renderer.js"
      >
        <div
          className="mosaic-window-toolbar draggable"
        >
          <div>
            <div>
              <h5>
                Renderer Process (renderer.js)
              </h5>
            </div>
            <div />
            <div
              className="mosaic-controls"
            >
              <button
                className="bp3-button bp3-small"
                onBlur={[Function]}
                onClick={[Function]}
                onKeyDown={[Function]}
                onKeyUp={[Function]}
                type="button"
              >
                <span
                  className="bp3-icon bp3-icon-maximize"
                  icon="maximize"
                >
                  <svg
                    data-icon="maximize"
                    height={16}
                    viewBox="0 0 16 16"
                    width={16}
                  >
                    <desc>
                      maximize
                    </desc>
                    <path
                      d="M5.99 8.99c-.28 0-.53.11-.71.29l-3.29 3.29v-1.59c0-.55-.45-1-1-1s-1 .45-1 1v4c0 .55.45 1 1 1h4c.55 0 1-.45 1-1s-.45-1-1-1H3.41L6.7 10.7a1.003 1.003 0 00-.71-1.71zm9-9h-4c-.55 0-1 .45-1 1s.45 1 1 1h1.59l-3.3 3.3a.99.99 0 00-.29.7 1.003 1.003 0 001.71.71l3.29-3.29V5c0 .55.45 1 1 1s1-.45 1-1V1c0-.56-.45-1.01-1-1.01z"
                      fillRule="evenodd"
                      key="0"
                    />
                  </svg>
                </span>
              </button>
              <button
                className="bp3-button bp3-small"
                onBlur={[Function]}
                onClick={[Function]}
                onKeyDown={[Function]}
                onKeyUp={[Function]}
                type="button"
              >
                <span
                  className="bp3-icon bp3-icon-cross"
                  icon="cross"
                >
                  <svg
                    data-icon="cross"
                    height={16}
                    viewBox="0 0 16 16"
                    width={16}
                  >
                    <desc>
                      cross
                    </desc>
                    <path
                      d="M9.41 8l3.29-3.29c.19-.18.3-.43.3-.71a1.003 1.003 0 00-1.71-.71L8 6.59l-3.29-3.3a1.003 1.003 0 00-1.42 1.42L6.59 8 3.3 11.29c-.19.18-.3.43-.3.71a1.003 1.003 0 001.71.71L8 9.41l3.29 3.29c.18.19.43.3.71.3a1.003 1.003 0 00.71-1.71L9.41 8z"
                      fillRule="evenodd"
                      key="0"
                    />
                  </svg>
                </span>
              </button>
            </div>
          </div>
        </div>
        <div
          className="mosaic-window-body"
        >
          Editor
        </div>
        <div
          className="mosaic-window-body-overlay"
          onClick={[Function]}
        />
        <div
          className="mosaic-window-additional-actions-bar"
        />
        <div
          className="mosaic-preview"
        >
          <div
            className="mosaic-window-toolbar"
          >
            <div
              className="mosaic-window-title"
            >
              Renderer Process (renderer.js)
            </div>
          </div>
          <div
            className="mosaic-window-body"
          >
            <h4>
              Renderer Process (renderer.js)
            </h4>
            <span
              className="bp3-icon bp3-icon-application"
              icon="application"
            >
              <svg
                data-icon="application"
                height={72}
                viewBox="0 0 20 20"
                width={72}
              >
                <desc>
                  application
                </desc>
                <path
                  d="M3.5 9h9c.28 0 .5-.22.5-.5s-.22-.5-.5-.5h-9c-.28 0-.5.22-.5.5s.22.5.5.5zm0 2h5c.28 0 .5-.22.5-.5s-.22-.5-.5-.5h-5c-.28 0-.5.22-.5.5s.22.5.5.5zM19 1H1c-.55 0-1 .45-1 1v16c0 .55.45 1 1 1h18c.55 0 1-.45 1-1V2c0-.55-.45-1-1-1zm-1 16H2V6h16v11zM3.5 13h7c.28 0 .5-.22.5-.5s-.22-.5-.5-.5h-7c-.28 0-.5.22-.5.5s.22.5.5.5z"
                  fillRule="evenodd"
                  key="0"
                />
              </svg>
            </span>
          </div>
        </div>
        <div
          className="drop-target-container"
        >
          <div
            className="drop-target top"
          />
          <div
            className="drop-target bottom"
          />
          <div
            className="drop-target left"
          />
          <div
            className="drop-target right"
          />
        </div>
      </div>
    </div>
    <div
      className="mosaic-split -row"
      onMouseDown={[Function]}
      style={
        Object {
          "bottom": "0%",
          "left": "50%",
          "right": "0%",
          "top": "0%",
        }
      }
    >
      <div
        className="mosaic-split-line"
      />
    </div>
    <div
      className="mosaic-tile"
      key="index.html"
      style={
        Object {
          "bottom": "50%",
          "left": "50%",
          "right": "0%",
          "top": "0%",
        }
      }
    >
      <div
        className="mosaic-window mosaic-drop-target index.html"
      >
        <div
          className="mosaic-window-toolbar draggable"
        >
          <div>
            <div>
              <h5>
                HTML (index.html)
              </h5>
            </div>
            <div />
            <div
              className="mosaic-controls"
            >
              <button
                className="bp3-button bp3-small"
                onBlur={[Function]}
                onClick={[Function]}
                onKeyDown={[Function]}
                onKeyUp={[Function]}
                type="button"
              >
                <span
                  className="bp3-icon bp3-icon-maximize"
                  icon="maximize"
                >
                  <svg
                    data-icon="maximize"
                    height={16}
                    viewBox="0 0 16 16"
                    width={16}
                  >
                    <desc>
                      maximize
                    </desc>
                    <path
                      d="M5.99 8.99c-.28 0-.53.11-.71.29l-3.29 3.29v-1.59c0-.55-.45-1-1-1s-1 .45-1 1v4c0 .55.45 1 1 1h4c.55 0 1-.45 1-1s-.45-1-1-1H3.41L6.7 10.7a1.003 1.003 0 00-.71-1.71zm9-9h-4c-.55 0-1 .45-1 1s.45 1 1 1h1.59l-3.3 3.3a.99.99 0 00-.29.7 1.003 1.003 0 001.71.71l3.29-3.29V5c0 .55.45 1 1 1s1-.45 1-1V1c0-.56-.45-1.01-1-1.01z"
                      fillRule="evenodd"
                      key="0"
                    />
                  </svg>
                </span>
              </button>
              <button
                className="bp3-button bp3-small"
                onBlur={[Function]}
                onClick={[Function]}
                onKeyDown={[Function]}
                onKeyUp={[Function]}
                type="button"
              >
                <span
                  className="bp3-icon bp3-icon-cross"
                  icon="cross"
                >
                  <svg
                    data-icon="cross"
                    height={16}
                    viewBox="0 0 16 16"
                    width={16}
                  >
                    <desc>
                      cross
                    </desc>
                    <path
                      d="M9.41 8l3.29-3.29c.19-.18.3-.43.3-.71a1.003 1.003 0 00-1.71-.71L8 6.59l-3.29-3.3a1.003 1.003 0 00-1.42 1.42L6.59 8 3.3 11.29c-.19.18-.3.43-.3.71a1.003 1.003 0 001.71.71L8 9.41l3.29 3.29c.18.19.43.3.71.3a1.003 1.003 0 00.71-1.71L9.41 8z"
                      fillRule="evenodd"
                      key="0"
                    />
                  </svg>
                </span>
              </button>
            </div>
          </div>
        </div>
        <div
          className="mosaic-window-body"
        >
          Editor
        </div>
        <div
          className="mosaic-window-body-overlay"
          onClick={[Function]}
        />
        <div
          className="mosaic-window-additional-actions-bar"
        />
        <div
          className="mosaic-preview"
        >
          <div
            className="mosaic-window-toolbar"
          >
            <div
              className="mosaic-window-title"
            >
              HTML (index.html)
            </div>
          </div>
          <div
            className="mosaic-window-body"
          >
            <h4>
              HTML (index.html)
            </h4>
            <span
              className="bp3-icon bp3-icon-application"
              icon="application"
            >
              <svg
                data-icon="application"
                height={72}
                viewBox="0 0 20 20"
                width={72}
              >
                <desc>
                  application
                </desc>
                <path
                  d="M3.5 9h9c.28 0 .5-.22.5-.5s-.22-.5-.5-.5h-9c-.28 0-.5.22-.5.5s.22.5.5.5zm0 2h5c.28 0 .5-.22.5-.5s-.22-.5-.5-.5h-5c-.28 0-.5.22-.5.5s.22.5.5.5zM19 1H1c-.55 0-1 .45-1 1v16c0 .55.45 1 1 1h18c.55 0 1-.45 1-1V2c0-.55-.45-1-1-1zm-1 16H2V6h16v11zM3.5 13h7c.28 0 .5-.22.5-.5s-.22-.5-.5-.5h-7c-.28 0-.5.22-.5.5s.22.5.5.5z"
                  fillRule="evenodd"
                  key="0"
                />
              </svg>
            </span>
          </div>
        </div>
        <div
          className="drop-target-container"
        >
          <div
            className="drop-target top"
          />
          <div
            className="drop-target bottom"
          />
          <div
            className="drop-target left"
          />
          <div
            className="drop-target right"
          />
        </div>
      </div>
    </div>
    <div
      className="mosaic-split -column"
      onMouseDown={[Function]}
      style={
        Object {
          "bottom": "0%",
          "left": "50%",
          "right": "0%",
          "top": "50%",
        }
      }
    >
      <div
        className="mosaic-split-line"
      />
    </div>
    <div
      className="mosaic-tile"
      key="preload.js"
      style={
        Object {
          "bottom": "25%",
          "left": "50%",
          "right": "0%",
          "top": "50%",
        }
      }
    >
      <div
        className="mosaic-window mosaic-drop-target preload.js"
      >
        <div
          className="mosaic-window-toolbar draggable"
        >
          <div>
            <div>
              <h5>
                Preload (preload.js)
              </h5>
            </div>
            <div />
            <div
              className="mosaic-controls"
            >
              <button
                className="bp3-button bp3-small"
                onBlur={[Function]}
                onClick={[Function]}
                onKeyDown={[Function]}
                onKeyUp={[Function]}
                type="button"
              >
                <span
                  className="bp3-icon bp3-icon-maximize"
                  icon="maximize"
                >
                  <svg
                    data-icon="maximize"
                    height={16}
                    viewBox="0 0 16 16"
                    width={16}
                  >
                    <desc>
                      maximize
                    </desc>
                    <path
                      d="M5.99 8.99c-.28 0-.53.11-.71.29l-3.29 3.29v-1.59c0-.55-.45-1-1-1s-1 .45-1 1v4c0 .55.45 1 1 1h4c.55 0 1-.45 1-1s-.45-1-1-1H3.41L6.7 10.7a1.003 1.003 0 00-.71-1.71zm9-9h-4c-.55 0-1 .45-1 1s.45 1 1 1h1.59l-3.3 3.3a.99.99 0 00-.29.7 1.003 1.003 0 001.71.71l3.29-3.29V5c0 .55.45 1 1 1s1-.45 1-1V1c0-.56-.45-1.01-1-1.01z"
                      fillRule="evenodd"
                      key="0"
                    />
                  </svg>
                </span>
              </button>
              <button
                className="bp3-button bp3-small"
                onBlur={[Function]}
                onClick={[Function]}
                onKeyDown={[Function]}
                onKeyUp={[Function]}
                type="button"
              >
                <span
                  className="bp3-icon bp3-icon-cross"
                  icon="cross"
                >
                  <svg
                    data-icon="cross"
                    height={16}
                    viewBox="0 0 16 16"
                    width={16}
                  >
                    <desc>
                      cross
                    </desc>
                    <path
                      d="M9.41 8l3.29-3.29c.19-.18.3-.43.3-.71a1.003 1.003 0 00-1.71-.71L8 6.59l-3.29-3.3a1.003 1.003 0 00-1.42 1.42L6.59 8 3.3 11.29c-.19.18-.3.43-.3.71a1.003 1.003 0 001.71.71L8 9.41l3.29 3.29c.18.19.43.3.71.3a1.003 1.003 0 00.71-1.71L9.41 8z"
                      fillRule="evenodd"
                      key="0"
                    />
                  </svg>
                </span>
              </button>
            </div>
          </div>
        </div>
        <div
          className="mosaic-window-body"
        >
          Editor
        </div>
        <div
          className="mosaic-window-body-overlay"
          onClick={[Function]}
        />
        <div
          className="mosaic-window-additional-actions-bar"
        />
        <div
          className="mosaic-preview"
        >
          <div
            className="mosaic-window-toolbar"
          >
            <div
              className="mosaic-window-title"
            >
              Preload (preload.js)
            </div>
          </div>
          <div
            className="mosaic-window-body"
          >
            <h4>
              Preload (preload.js)
            </h4>
            <span
              className="bp3-icon bp3-icon-application"
              icon="application"
            >
              <svg
                data-icon="application"
                height={72}
                viewBox="0 0 20 20"
                width={72}
              >
                <desc>
                  application
                </desc>
                <path
                  d="M3.5 9h9c.28 0 .5-.22.5-.5s-.22-.5-.5-.5h-9c-.28 0-.5.22-.5.5s.22.5.5.5zm0 2h5c.28 0 .5-.22.5-.5s-.22-.5-.5-.5h-5c-.28 0-.5.22-.5.5s.22.5.5.5zM19 1H1c-.55 0-1 .45-1 1v16c0 .55.45 1 1 1h18c.55 0 1-.45 1-1V2c0-.55-.45-1-1-1zm-1 16H2V6h16v11zM3.5 13h7c.28 0 .5-.22.5-.5s-.22-.5-.5-.5h-7c-.28 0-.5.22-.5.5s.22.5.5.5z"
                  fillRule="evenodd"
                  key="0"
                />
              </svg>
            </span>
          </div>
        </div>
        <div
          className="drop-target-container"
        >
          <div
            className="drop-target top"
          />
          <div
            className="drop-target bottom"
          />
          <div
            className="drop-target left"
          />
          <div
            className="drop-target right"
          />
        </div>
      </div>
    </div>
    <div
      className="mosaic-split -column"
      onMouseDown={[Function]}
      style={
        Object {
          "bottom": "0%",
          "left": "50%",
          "right": "0%",
          "top": "75%",
        }
      }
    >
      <div
        className="mosaic-split-line"
      />
    </div>
    <div
      className="mosaic-tile"
      key="styles.css"
      style={
        Object {
          "bottom": "0%",
          "left": "50%",
          "right": "0%",
          "top": "75%",
        }
      }
    >
      <div
        className="mosaic-window mosaic-drop-target styles.css"
      >
        <div
          className="mosaic-window-toolbar draggable"
        >
          <div>
            <div>
              <h5>
                Stylesheet (styles.css)
              </h5>
            </div>
            <div />
            <div
              className="mosaic-controls"
            >
              <button
                className="bp3-button bp3-small"
                onBlur={[Function]}
                onClick={[Function]}
                onKeyDown={[Function]}
                onKeyUp={[Function]}
                type="button"
              >
                <span
                  className="bp3-icon bp3-icon-maximize"
                  icon="maximize"
                >
                  <svg
                    data-icon="maximize"
                    height={16}
                    viewBox="0 0 16 16"
                    width={16}
                  >
                    <desc>
                      maximize
                    </desc>
                    <path
                      d="M5.99 8.99c-.28 0-.53.11-.71.29l-3.29 3.29v-1.59c0-.55-.45-1-1-1s-1 .45-1 1v4c0 .55.45 1 1 1h4c.55 0 1-.45 1-1s-.45-1-1-1H3.41L6.7 10.7a1.003 1.003 0 00-.71-1.71zm9-9h-4c-.55 0-1 .45-1 1s.45 1 1 1h1.59l-3.3 3.3a.99.99 0 00-.29.7 1.003 1.003 0 001.71.71l3.29-3.29V5c0 .55.45 1 1 1s1-.45 1-1V1c0-.56-.45-1.01-1-1.01z"
                      fillRule="evenodd"
                      key="0"
                    />
                  </svg>
                </span>
              </button>
              <button
                className="bp3-button bp3-small"
                onBlur={[Function]}
                onClick={[Function]}
                onKeyDown={[Function]}
                onKeyUp={[Function]}
                type="button"
              >
                <span
                  className="bp3-icon bp3-icon-cross"
                  icon="cross"
                >
                  <svg
                    data-icon="cross"
                    height={16}
                    viewBox="0 0 16 16"
                    width={16}
                  >
                    <desc>
                      cross
                    </desc>
                    <path
                      d="M9.41 8l3.29-3.29c.19-.18.3-.43.3-.71a1.003 1.003 0 00-1.71-.71L8 6.59l-3.29-3.3a1.003 1.003 0 00-1.42 1.42L6.59 8 3.3 11.29c-.19.18-.3.43-.3.71a1.003 1.003 0 001.71.71L8 9.41l3.29 3.29c.18.19.43.3.71.3a1.003 1.003 0 00.71-1.71L9.41 8z"
                      fillRule="evenodd"
                      key="0"
                    />
                  </svg>
                </span>
              </button>
            </div>
          </div>
        </div>
        <div
          className="mosaic-window-body"
        >
          Editor
        </div>
        <div
          className="mosaic-window-body-overlay"
          onClick={[Function]}
        />
        <div
          className="mosaic-window-additional-actions-bar"
        />
        <div
          className="mosaic-preview"
        >
          <div
            className="mosaic-window-toolbar"
          >
            <div
              className="mosaic-window-title"
            >
              Stylesheet (styles.css)
            </div>
          </div>
          <div
            className="mosaic-window-body"
          >
            <h4>
              Stylesheet (styles.css)
            </h4>
            <span
              className="bp3-icon bp3-icon-application"
              icon="application"
            >
              <svg
                data-icon="application"
                height={72}
                viewBox="0 0 20 20"
                width={72}
              >
                <desc>
                  application
                </desc>
                <path
                  d="M3.5 9h9c.28 0 .5-.22.5-.5s-.22-.5-.5-.5h-9c-.28 0-.5.22-.5.5s.22.5.5.5zm0 2h5c.28 0 .5-.22.5-.5s-.22-.5-.5-.5h-5c-.28 0-.5.22-.5.5s.22.5.5.5zM19 1H1c-.55 0-1 .45-1 1v16c0 .55.45 1 1 1h18c.55 0 1-.45 1-1V2c0-.55-.45-1-1-1zm-1 16H2V6h16v11zM3.5 13h7c.28 0 .5-.22.5-.5s-.22-.5-.5-.5h-7c-.28 0-.5.22-.5.5s.22.5.5.5z"
                  fillRule="evenodd"
                  key="0"
                />
              </svg>
            </span>
          </div>
        </div>
        <div
          className="drop-target-container"
        >
          <div
            className="drop-target top"
          />
          <div
            className="drop-target bottom"
          />
          <div
            className="drop-target left"
          />
          <div
            className="drop-target right"
          />
        </div>
      </div>
    </div>
  </div>
  <div
    className="drop-target-container"
  >
    <div
      className="drop-target top"
    />
    <div
      className="drop-target bottom"
    />
    <div
      className="drop-target left"
    />
    <div
      className="drop-target right"
    />
  </div>
</div>
`;

exports[`Editors component renders a toolbar 1`] = `
<div>
  <div>
    <h5>
      main.js
    </h5>
  </div>
  <div />
  <div
    className="mosaic-controls"
  >
    <MaximizeButton
      appState={
        Object {
          "editorMosaic": Object {
            "layoutDebounce": 26,
            "mosaic": Object {
              "direction": "row",
              "first": Object {
                "direction": "column",
                "first": "main.js",
                "second": "renderer.js",
              },
              "second": Object {
                "direction": "column",
                "first": "index.html",
                "second": Object {
                  "direction": "column",
                  "first": "preload.js",
                  "second": "styles.css",
                },
              },
            },
          },
        }
      }
      id="main.js"
    />
    <RemoveButton
      appState={
        Object {
          "editorMosaic": Object {
            "layoutDebounce": 26,
            "mosaic": Object {
              "direction": "row",
              "first": Object {
                "direction": "column",
                "first": "main.js",
                "second": "renderer.js",
              },
              "second": Object {
                "direction": "column",
                "first": "index.html",
                "second": Object {
                  "direction": "column",
                  "first": "preload.js",
                  "second": "styles.css",
                },
              },
            },
          },
        }
      }
      id="main.js"
    />
  </div>
</div>
`;
